### DISCUSSION TITLE
How to detect actual string type

### DISCUSSION THREAD

**Discussion by FBruzzesi** [dask/dask#11297] [Q&A]
## Description

I cannot find a way to assess if the underlying data is an actual string or not.

Consider the following snippet:

```py
import dask.dataframe as dd

class Foo:
    pass
    
data = {
    "a": list("xyz"),
    "b": [[1], [2, 3], [3, 2, 1]],
    "c": [Foo(), Foo, Foo()]
}

df = dd.from_dict(data, npartitions=1)
df.dtypes
```
> a    string[pyarrow]
> b    string[pyarrow]
> c    string[pyarrow]
> dtype: object

and (maybe even worse), if I try running any `.str` namespace method, I get no issue/warning/raise:

```py
df["c"].str.split().compute()
```
Results in 

> 0    [<__main__.Foo, object, at, 0x7fd9a53b75e0>]
> 1                       [<class, '__main__.Foo'>]
> 2    [<__main__.Foo, object, at, 0x7fd9a53b75b0>]
> Name: c, dtype: object

## Pandas counterpart

In pandas land, some tricks can be applied, depending on how certain we want to be that a column is backed by actual strings.

Example:
```py
import pandas as pd

df = pd.DataFrame(data)
isinstance(df["c"].loc[df["c"].first_valid_index()], str)
```
> False

In general I can get a bunch of random indexes and check if they are strings or not.

## Additional info

dask==2024.6.2
pandas==2.2.2
pyarrow==17.0.0

**Comment by TayfurYldz** [comment]
What you're seeing is Dask's automatic object/string conversion, so the resulting dtype is not evidence that every original Python object was a `str`.

On current Dask this behavior is controlled by `dataframe.convert-string`; object-like input can be converted to a PyArrow-backed string dtype. If preserving the original Python objects is what you need, disable that conversion when constructing/reading the collection:

```python
import dask

with dask.config.set({"dataframe.convert-string": False}):
    df = dd.from_dict(data, npartitions=1)
```

Then the object columns remain object-typed instead of being stringified by the Arrow-string conversion step.

There is a separate issue in the question, though: an `object` dtype still cannot prove that all values are strings. Pandas has the same fundamental limitation because an object column may contain heterogeneous Python values. If you need a data-quality guarantee ("every non-null value is really a Python str"), that has to be validated from the values, preferably partition-wise rather than by sampling a few random rows. For example, map a predicate over partitions and reduce it to a single boolean.

So I would use the dtype for storage/schema information, and an explicit validation for the stronger semantic invariant. Sampling the first/random value can miss mixed-type partitions.
